### PR TITLE
TextInput placeholder should uses a theme color

### PR DESCRIPTION
**Pull request by Nitin-100**
## Description
The TextInput placeholder color is now theme-aware and adapts intelligently based on background color and focus state, replacing the previous hardcoded gray color (D2D1::ColorF::Gray).
The implementation uses ITU-R BT.601 luminance calculation to determine background brightness and automatically selects appropriate placeholder colors that provide optimal contrast and visibility across different themes and backgrounds.

##Bug
https://github.com/microsoft/react-native-windows/issues/15129

### Why
The hardcoded gray placeholder color (D2D1::ColorF::Gray) provided poor accessibility and visual consistency across different Windows themes and custom background colors. Users experienced:
-Poor contrast on dark themes
-Invisible or barely visible placeholders on certain background colors
-No differentiation between focused and unfocused states
-Inconsistent visual experience across Light/Dark Windows themes

This fix ensures placeholder colors adapt intelligently to provide optimal contrast and accessibility across all scenarios.

### What
-Replaced hardcoded D2D1::ColorF::Gray with intelligent color selection
-Added GetTextInputPlaceholderColor() function that considers:
       1.Background color brightness using ITU-R BT.601 luminance calculation
       2.Focus state (focused vs unfocused TextInput)
       3.High contrast mode for accessibility compliance
       4.Uses existing IsColorLight() utility from ThemeUtils for consistent color logic
       5.Provides fallback to WinUI3 semantic colors when no background is specified
       6.Light backgrounds get darker placeholder colors for better contrast
       7.Dark backgrounds get lighter placeholder colors for better visibility
       8.Focused inputs get higher contrast colors than unfocused inputs

## Testing

https://github.com/user-attachments/assets/c854c763-0b81-466f-bd00-e4cede43d8a2

Added below textinputs.tsx for testing 
            {/* THEME-AWARE PLACEHOLDER TESTING HEADER */}
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#e3f2fd',
                  color: 'black',
                  borderColor: 'blue',
                  borderWidth: 3,
                  height: 100,
                },
              ]}
              placeholder="THEME-AWARE PLACEHOLDER TESTING: Uses ITU-R BT.601 luminance calculation for background brightness detection. Try: 1) Changing Windows theme (Light/Dark) 2) Enabling High Contrast mode 3) Clicking inputs to test focus states"
              multiline={true}
              editable={false}
            />

            {/* =========================== */}
            {/* THEME-AWARE PLACEHOLDER TESTS */}
            {/* =========================== */}

            {/* Test 1: Default System Theme Adaptation */}
            <TextInput
              style={[styles.input, {borderColor: 'blue', borderWidth: 3}]}
              placeholder=" System Theme - Default background (focus to see difference)"
            />

            {/* Test 2: Light Background Tests */}
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: 'white',
                  color: 'black',
                  borderColor: 'green',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Light Background - Should show dark placeholder"
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#f5f5f5',
                  color: 'black',
                  borderColor: 'green',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Light Gray Background - Focus to see darker placeholder"
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#e8f4fd',
                  color: 'black',
                  borderColor: 'green',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Light Blue Background - Test contrast adaptation"
            />

            {/* Test 3: Dark Background Tests */}
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: 'black',
                  color: 'white',
                  borderColor: 'yellow',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Dark Background - Should show light placeholder"
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#2d2d2d',
                  color: 'white',
                  borderColor: 'yellow',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Dark Gray Background - Focus to see lighter placeholder"
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#1a1a2e',
                  color: 'white',
                  borderColor: 'yellow',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Dark Blue Background - Test contrast adaptation"
            />

            {/* Test 4: Medium Tone Background Tests */}
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#808080',
                  color: 'white',
                  borderColor: 'orange',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Medium Gray Background - Brightness threshold test"
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#4a4a4a',
                  color: 'white',
                  borderColor: 'orange',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Medium Dark Background - Edge case test"
            />

            {/* Test 5: Colored Background Tests */}
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#ffebee',
                  color: 'black',
                  borderColor: 'red',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Light Red Background - Color brightness test"
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#b71c1c',
                  color: 'white',
                  borderColor: 'red',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Dark Red Background - Color brightness test"
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#e8f5e8',
                  color: 'black',
                  borderColor: 'green',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Light Green Background - Color brightness test"
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#1b5e20',
                  color: 'white',
                  borderColor: 'green',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Dark Green Background - Color brightness test"
            />

            {/* Test 6: High Contrast Simulation */}
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: 'white',
                  color: 'black',
                  borderColor: 'black',
                  borderWidth: 3,
                },
              ]}
              placeholder=" High Contrast Simulation - White BG"
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: 'black',
                  color: 'white',
                  borderColor: 'white',
                  borderWidth: 3,
                },
              ]}
              placeholder=" High Contrast Simulation - Black BG"
            />

            {/* Test 7: Transparent/No Background */}
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: 'transparent',
                  borderColor: 'purple',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Transparent Background - Fallback test"
            />

            {/* Test 8: Edge Cases */}
            <TextInput

              style={[
                styles.input,
                {
                  backgroundColor: '#ffffff00',
                  borderColor: 'gray',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Fully Transparent Background - Alpha test"
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#00000000',
                  borderColor: 'gray',
                  borderWidth: 2,
                },
              ]}
              placeholder=" Transparent Black Background - Alpha test"
            />

            {/* Test 9: Focus State Comparison */}
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: 'white',
                  color: 'black',
                  borderColor: 'blue',
                  borderWidth: 2,
                },
              ]}
              placeholder=" CLICK HERE: Light bg - compare focused vs unfocused"
              autoFocus={false}
            />
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: 'black',
                  color: 'white',
                  borderColor: 'cyan',
                  borderWidth: 2,
                },
              ]}
              placeholder=" CLICK HERE: Dark bg - compare focused vs unfocused"
              autoFocus={false}
            />

            {/* Test 10: Auto-focused for immediate visibility */}
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#f0f0f0',
                  color: 'black',
                  borderColor: 'red',
                  borderWidth: 3,
                },
              ]}
              placeholder=" AUTO-FOCUSED: Light bg showing focused placeholder color"
              autoFocus={true}
            />

            {/* Instructions */}
            <TextInput
              style={[
                styles.input,
                {
                  backgroundColor: '#ffffcc',
                  color: 'black',
                  borderColor: 'gold',
                  borderWidth: 2,
                  height: 80,
                },
              ]}
              placeholder="TESTING INSTRUCTIONS: 1) Change Windows theme (Settings > Personalization > Colors) 2) Test High Contrast mode 3) Click different inputs to see focus changes 4) Compare placeholder visibility on different backgrounds"
              multiline={true}
              editable={false}
            />

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/react-native-windows/pull/15161)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp [hunk 1]
```diff
@@ -142,4 +144,35 @@ winrt::Windows::UI::Color ResolvePlatformColor(const std::vector<std::string> &s
   return {};
 }
 
+winrt::Windows::UI::Color GetTextInputPlaceholderColor(
+    bool isFocused,
+    const winrt::Windows::UI::Color &backgroundColor) {
+  // In high contrast mode, always use system GrayText for accessibility
+  HIGHCONTRAST hcInfo = {};
+  if (SystemParametersInfo(SPI_GETHIGHCONTRAST, 0, &hcInfo, 0) && (hcInfo.dwFlags & HCF_HIGHCONTRASTON)) {
```
Comment (chrisglein) [inline vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp]: I'd be shocked if this was the only "is high contrast" check that ends up persisting in PlatformColor. Worth making into a clear helper method and a clear legible boolean result.

File: vnext/Microsoft.ReactNative/Fabric/Composition/TextInput/WindowsTextInputComponentView.cpp [hunk 2]
```diff
@@ -1690,8 +1691,20 @@ void WindowsTextInputComponentView::DrawText() noexcept {
...
+            auto bgColor = (*props.backgroundColor).AsWindowsColor();
+            backgroundColor = bgColor;
+          }
+
+          auto placeholderColor = facebook::react::GetTextInputPlaceholderColor(m_hasFocus, backgroundColor);
+          auto d2dColor = D2D1::ColorF(
+              placeholderColor.R / 255.0f,
+              placeholderColor.G / 255.0f,
+              placeholderColor.B / 255.0f,
+              placeholderColor.A / 255.0f);
```
Comment (chrisglein) [inline vnext/Microsoft.ReactNative/Fabric/Composition/TextInput/WindowsTextInputComponentView.cpp]: There should be a standard conversion helper for this. Might be Theme::D2DColor? Something like that.

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp [hunk 3]
```diff
@@ -142,4 +144,35 @@ winrt::Windows::UI::Color ResolvePlatformColor(const std::vector<std::string> &s
...
+  // Use ITU-R BT.601 luminance calculation to determine background brightness
+  bool isLightBackground = Microsoft::ReactNative::IsColorLight(backgroundColor);
+
+  if (isLightBackground) {
+    return isFocused ? winrt::Windows::UI::Color{0xFF, 0x60, 0x60, 0x60} : // Darker gray for focused state
+        winrt::Windows::UI::Color{0xFF, 0x99, 0x99, 0x99}; // Medium gray for unfocused state
+  } else {
+    return isFocused ? winrt::Windows::UI::Color{0xFF, 0xC0, 0xC0, 0xC0} : // Light gray for focused state
+        winrt::Windows::UI::Color{0xFF, 0x90, 0x90, 0x90}; // Medium gray for unfocused state
+  }
```
Comment (chrisglein) [inline vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp:189]: As done above, can you attribute to the source colors in XAML, assuming you matched what XAML is doing here.

**Comment by chrisglein** [comment]
@Nitin-100 

> Type of Change
_Bug fix (non-breaking change which fixes an issue)

No linked bug #?

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp (hunk 1 above)
Comment (Nitin-100) [reply vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp]: Updated with an API already implemented.

File: vnext/Microsoft.ReactNative/Fabric/Composition/TextInput/WindowsTextInputComponentView.cpp (hunk 2 above)
Comment (Nitin-100) [reply vnext/Microsoft.ReactNative/Fabric/Composition/TextInput/WindowsTextInputComponentView.cpp]: Updated with the standard conversion API.

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp (hunk 3 above)
Comment (Nitin-100) [reply vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp:189]: Updated with Platform specified colors.

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.h [hunk 4]
```diff
@@ -5,8 +5,20 @@
...
+#ifdef USE_FABRIC
+class SharedColor;
+#endif
+}
+
 namespace facebook::react {
 
 winrt::Windows::UI::Color ResolvePlatformColor(const std::vector<std::string> &semanticItems);
 
+#ifdef USE_FABRIC
```
Comment (iamAbhi-916) [inline vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.h]: no need of #ifdef USE_FABRIC , this file is already Fabric specific .

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.h [hunk 5]
```diff
@@ -5,8 +5,20 @@
 
 #include <winrt/Windows.UI.h>
 
+// Forward declaration to avoid circular dependencies
+namespace facebook::react {
+#ifdef USE_FABRIC
```
Comment (iamAbhi-916) [inline vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.h]: no need of #ifdef USE_FABRIC , this file is already Fabric specific .

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp [hunk 6]
```diff
@@ -142,4 +147,50 @@ winrt::Windows::UI::Color ResolvePlatformColor(const std::vector<std::string> &s
   return {};
 }
 
+#ifdef USE_FABRIC
```
Comment (iamAbhi-916) [inline vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp]: no need of #ifdef USE_FABRIC , this file is already Fabric specific .

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp [hunk 7]
```diff
@@ -142,4 +147,50 @@ winrt::Windows::UI::Color ResolvePlatformColor(const std::vector<std::string> &s
...
+SharedColor GetTextInputPlaceholderColor(bool isFocused, const winrt::Windows::UI::Color &backgroundColor) {
+  // In high contrast mode, always use system GrayText for accessibility
+  if (Microsoft::ReactNative::IsInHighContrastWin32()) {
```
Comment (iamAbhi-916) [inline vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp]: can we implement it using winrt (winrt::Windows::UI::ViewManagement::AccessibilitySettings()) instead of win32 (suggestion)

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp [hunk 8]
```diff
@@ -142,4 +147,50 @@ winrt::Windows::UI::Color ResolvePlatformColor(const std::vector<std::string> &s
...
+  // If no background color provided, use WinUI3 semantic colors as fallback
+  if (backgroundColor.A == 0) {
+    if (isFocused) {
+      auto color = ResolvePlatformColor({"TextFillColorSecondary"});
+      return hostPlatformColorFromRGBA(color.R, color.G, color.B, color.A);
+    } else {
+      auto color = ResolvePlatformColor({"TextFillColorDisabled"});
+      return hostPlatformColorFromRGBA(color.R, color.G, color.B, color.A);
+    }
+  }
```
Comment (chrisglein) [inline vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp]: Does this do what I'm thinking, as in add a background color TextInputs that did not supply them? How does this compare with RN core on iOS/Andoid? I recall us making an intentional decision with Fabric to match RN core and not RNW Paper.

It's a policy decision as far as which we match. Not matching Paper means more work for those migrating (and thus documentation). Not matching RN core iOS/Android means more confusion when developing cross-platform apps.

Comment (Nitin-100) [reply vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp:168]: The intent here is to determine the **placeholder TEXT color** (not background), but I see how checking `backgroundColor.A == 0` could be misleading. When a TextInput has no background color specified, I was trying to provide sensible placeholder text colors, but this shouldn't affect the TextInput's background itself.

Let me research how iOS/Android handle placeholder text colors for TextInputs with transparent/no backgrounds and align with that behavior. Would you prefer we match RN core iOS/Android behavior here, or should we maintain RNW Paper compatibility?

Comment (Nitin-100) [reply vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp]: After researching iOS/Android defaults, I found that React Native core relies on "platform-provided default colors" when no background is specified, rather than implementing custom logic like mine was doing.

I've updated the implementation to:
1. Match RN Core behavior: When `backgroundColor.A == 0` (transparent/no background), use Windows system defaults (`WindowText` at 60% opacity) rather than custom WinUI3 semantic colors
2. Remove focus-dependent logic for the transparent case - iOS/Android don't change placeholder color based on focus state
3. Maintain the background-aware contrast logic for when an actual background color is provided

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp (hunk 7 above)
Comment (Nitin-100) [reply vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.cpp]: Yeah, more consistent winrt call. Done.

File: vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.h (hunk 4 above)
Comment (Nitin-100) [reply vnext/Microsoft.ReactNative/Fabric/platform/react/renderer/graphics/PlatformColorUtils.h]: Done.

**Review by chrisglein — APPROVED** [review]
